pkg/drivers/krunkit/krunkit.go

```diff
@@ -0,0 +1,561 @@
+		exit.Error(reason.IfBootpdFirewall, "ip not found", err)
+	}
+	out.Styled(style.Restarting, "Successfully unblocked bootpd process from firewall, retrying")
+	return fmt.Errorf("ip not found: %v", err)
+}
+
+func isBootpdError(err error) bool {
+	return strings.Contains(err.Error(), "could not find an IP address")
+}
+
+func (d *Driver) Stop() error {
+	if err := d.stopKrunkit(); err != nil {
+		return err
+	}
+	return d.VmnetHelper.Stop()
+}
+
+func (d *Driver) Kill() error {
+	if err := d.killkrunkit(); err != nil {
+		return err
+	}
+	return d.VmnetHelper.Kill()
+}
+
+func (d *Driver) Remove() error {
+	s, err := d.GetState()
+	if err != nil {
+		return errors.Wrap(err, "get state")
+	}
+	if s == state.Running {
+		if err := d.Kill(); err != nil {
+			return errors.Wrap(err, "kill")
+		}
+	}
+	return nil
+}
+
+func (d *Driver) Restart() error {
+	s, err := d.GetState()
+	if err != nil {
+		return err
+	}
+
+	if s == state.Running {
+		if err := d.Stop(); err != nil {
+			return err
+		}
+	}
+	return d.Start()
+}
+
+func (d *Driver) StartDocker() error {
+	return fmt.Errorf("hosts without a driver cannot start docker")
+}
+
+func (d *Driver) StopDocker() error {
+	return fmt.Errorf("hosts without a driver cannot stop docker")
+}
+
+func (d *Driver) GetDockerConfigDir() string {
+	return ""
+}
+
+func (d *Driver) Upgrade() error {
+	return fmt.Errorf("hosts without a driver cannot be upgraded")
+}
+
+func (d *Driver) getKrunkitState() (state.State, error) {
+	pidfile := d.pidfilePath()
+	pid, err := process.ReadPidfile(pidfile)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return state.Error, err
+		}
+		return state.Stopped, nil
+	}
+	exists, err := process.Exists(pid, driverName)
+	if err != nil {
+		return state.Error, err
+	}
+	if !exists {
+		// No process, stale pidfile.
+		if err := os.Remove(pidfile); err != nil {
+			log.Debugf("failed to remove %q: %s", pidfile, err)
+		}
+		return state.Stopped, nil
+	}
+	return state.Running, nil
+}
+
+func (d *Driver) stopKrunkit() error {
+	if err := d.SetKrunkititState("Stop"); err != nil {
+		// krunkit may be already stopped, shutting down, or not listening.
+		log.Debugf("Failed to set krunkit state to 'Stop': %s", err)
+		pidfile := d.pidfilePath()
+		pid, err := process.ReadPidfile(pidfile)
+		if err != nil {
+			if !errors.Is(err, os.ErrNotExist) {
+				return err
+			}
+			// No pidfile.
+			return nil
+		}
+		if err := process.Terminate(pid, driverName); err != nil {
+			if err != os.ErrProcessDone {
+				return err
+			}
+			// No process, stale pidfile.
+			if err := os.Remove(pidfile); err != nil {
+				log.Debugf("failed to remove %q: %s", pidfile, err)
+			}
+			return nil
+		}
+	}
+	return nil
+}
+
+func (d *Driver) killkrunkit() error {
+	if err := d.SetKrunkititState("HardStop"); err != nil {
+		// Typically fails with EOF due to https://github.com/crc-org/krunkit/issues/277.
+		log.Debugf("Failed to set krunkit state to 'HardStop': %s", err)
```

Comment:
krunkit ignores the value and does not document this value, so we should not set the state here but just kill it.
https://github.com/containers/krunkit/blob/9421c0e1daca8af0eb84d6afc67bfb55b78214c4/src/status.rs#L101

Reply:
Fixed in current version